File: Java/src/test/java/org/codecop/lcdnumbers/DigitPrinterTest.java
package org.codecop.lcdnumbers;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class DigitPrinterTest {

    final DigitPrinter printer = new DigitPrinter(new Zipper());
    // Zipper feels technical, so it is rather a helper than a real collaborator.

    @Test
    public void shouldRenderEmptyDigit() {
        // maybe it should not because it does not make sense for business
        String display = printer.render(Collections.emptyList());

        assertEquals("\n", display);
    }

    @Test
    public void shouldRenderDigitsSideBySide() {
        Digit one = from("A1", //
                         "A2");
        Digit two = from("B1", //
                         "B2");

        String display = printer.render(Arrays.asList(one, two));

        assertEquals("A1B1\n" + //
                     "A2B2\n", display);
    }

    private Digit from(String line1, String line2) {
        Digit digit = mock(Digit.class);
        List<Line> lines = Arrays.asList(new Line(line1), new Line(line2));
        when(digit.lines()).thenReturn(lines);
        return digit;
    }

}

File: JavaScript/test/digit_printer_test.js
/* globals describe, beforeEach, it, zip, DigitPrinter */
'use strict';

var chai = require('chai');
var should = chai.should();

require('../src/line');
require('../src/digit_printer');
require('../src/zip');

function digitFrom(line1, line2) {
    var lines = [ new Line(line1), new Line(line2) ];
    return { lines: function() { return lines; }}; // stub Digit
}

describe('DigitPrinter', function () {

    var printer;

    beforeEach(function() {
        // Zip feels technical, so it is rather a helper than a real collaborator.
        printer = new DigitPrinter(zip);
    });

    it('should render digits side by side', function () {
        var one = digitFrom('A1', //
                            'A2');
        var two = digitFrom('B1', //
                            'B2');

        var display = printer.render([ one, two ]);

        display.should.equal('A1B1\n' + //
                             'A2B2\n');
    });

});
